bridgebot/bidding.py

```diff
@@ -0,0 +1,158 @@
+from enums import Players, Strains, AuctionStatus, Doubles, Contracts, PASS, Team
+
+
+class InvalidDealerException(Exception):
+    pass
+
+
+class Contract:
+    def __init__(self):
+        self.level = 0
+        self.strain = Strains.NT
+        self.doubled = Doubles.NONE
+        self.declarer = None
+
+
+class Auction:
+    def __init__(self, dealer):
+        self.dealer = dealer.upper()
+
+        if not isinstance(dealer, Players):
+            raise InvalidDealerException('Invalid dealer')
+
+        self.record = {
+            Players.NORTH: [],
+            Players.EAST: [],
+            Players.SOUTH: [],
+            Players.WEST: []
+        }
+
+        self.ns_first_bid = {
+            Strains.CLUBS: None,
+            Strains.DIAMONDS: None,
+            Strains.HEARTS: None,
+            Strains.SPADES: None,
+            Strains.NT: None
+        }
+
+        self.ew_first_bid = {
+            Strains.CLUBS: None,
+            Strains.DIAMONDS: None,
+            Strains.HEARTS: None,
+            Strains.SPADES: None,
+            Strains.NT: None
+        }
+
+        self.player_index = Players.players().index(self.dealer)
+        self.last_bidder_index = None
+
+        # Not sure about the following line
+        for player in Players.players():
+            self.record[player].append("-")
+
+        self.bid_index = -1
+        self.doubled_by = None
+        self.redoubled = False
+        self.consecutive_passes = 0
+
+    # For debugging purposes, easier to reset than to create a new Auction
+    def reset(self):
+        contract = Contract()
+
+        self.player_index = players.index(self.dealer)
+        self.last_bidder_index = None
+
+        for i in range(self.player_index):
+            self.record[players[i]].append("-")
+
+        self.bid_index = -1
+        self.doubled_by = None # Redundant with contract dictionary, but makes the logic below much cleaner.
+        self.redoubled = False
+        self.consecutive_passes = 0
+
+    def __increment_player(self):
+        self.player = self.player.next_player()
+
+    def pass_bid(self):
+        if self.bid_index == -1:
+            if self.consecutive_passes == 3: # Passing out
+                self.record[self.player].append(PASS)
+                self.contract.strain = Strains.NT
+                return AuctionStatus.DONE
+            else:  # Passing before an opening bid
+                self.record[self.player].append(PASS)
+                self.__increment_player()
+                return AuctionStatus.CONTINUE
+        elif self.consecutive_passes != 2: # Passing but not finishing
+            self.record[self.player].append(PASS)
+            self.__increment_player()
+            return AuctionStatus.CONTINUE
+        else:  # Three passes in a row.
+            if self.redoubled:
+                self.contract.doubled = Doubles.DOUBLE_DOWN
```

Comment:
`DOUBLE_DOWN` ?

Reply:
If X is Double, XX is Double Down. Mable a better name is DOUBLE_DOUBLE or QUADRUPLE